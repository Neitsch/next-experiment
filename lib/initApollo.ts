import { InMemoryCache } from "apollo-cache-inmemory";
import { ApolloClient } from "apollo-client";
import { setContext } from "apollo-link-context";
import { HttpLink } from "apollo-link-http";
import { createPersistedQueryLink } from "apollo-link-persisted-queries";
import isomorphicUnfetch from "isomorphic-unfetch";

let apolloClient = null;

// Polyfill fetch() on the server (used by apollo-client)
// if (!process.browser) {
//   global.fetch = fetch
// }

export const setAuthorizationLink = jwt =>
  setContext(() => ({
    headers: {
      authorization: "Bearer " + jwt,
    },
  }));

function create(initialState, jwt) {
  return new ApolloClient({
    cache: new InMemoryCache().restore(initialState || {}),
    connectToDevTools: !!process.browser,
    link: setAuthorizationLink(jwt).concat(
      createPersistedQueryLink().concat(
        new HttpLink({
          credentials: "same-origin",
          fetch: isomorphicUnfetch,
          uri: "http://localhost:3000/graphql",
        }),
      ),
    ),
    ssrMode: !process.browser,
  });
}

export const resetClient = () => {
  apolloClient = null;
};

export default function initApollo(initialState, jwt) {
  if (!process.browser) {
    return create(initialState, jwt);
  }
  if (!apolloClient) {
    apolloClient = create(initialState, jwt);
  }
  return apolloClient;
}
