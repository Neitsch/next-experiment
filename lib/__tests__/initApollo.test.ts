import InitApollo, { resetClient, setAuthorizationLink } from "../initApollo";

describe("Init Apollo", () => {
  beforeEach(() => {
    resetClient();
  });

  describe("Server", () => {
    it("Creates - no data", () => {
      process.browser = false;
      const apollo = InitApollo(null);
      expect(apollo.cache.data.data).toEqual({});
      expect(apollo.ssrMode).toBe(true);
    });
    it("Creates - data", () => {
      process.browser = false;
      const data = {
        testData: "Hi",
      };
      const apollo = InitApollo(data);
      expect(apollo.cache.data.data).toEqual(data);
      expect(apollo.ssrMode).toBe(true);
    });
  });

  describe("Browser", () => {
    it("Creates - no data", () => {
      process.browser = true;
      const apollo = InitApollo(null);
      expect(apollo.cache.data.data).toEqual({});
      expect(apollo.ssrMode).toBe(false);
    });
    it("Creates - data", () => {
      process.browser = true;
      const data = {
        testData: "Hi",
      };
      const apollo = InitApollo(data);
      expect(apollo.cache.data.data).toEqual(data);
      expect(apollo.ssrMode).toBe(false);
    });
    it("Creates - no duplicate", () => {
      process.browser = true;
      const apollo1 = InitApollo({});
      const apollo2 = InitApollo({});
      expect(Object.is(apollo1, apollo2)).toBe(true);
    });
  });

  it("Auth link", () => {
    jest.resetAllMocks();
    jest.resetModules();
    const setContext = jest.fn();
    jest.doMock("apollo-link-context", () => ({
      setContext,
    }));
    require("../initApollo").setAuthorizationLink("My jwt");
    expect(setContext.mock.calls[0][0]()).toMatchSnapshot();
  });
});
