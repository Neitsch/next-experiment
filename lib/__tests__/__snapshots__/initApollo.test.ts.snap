// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init Apollo Auth link 1`] = `
Object {
  "headers": Object {
    "authorization": "Bearer My jwt",
  },
}
`;
